pkg/ops/rawDiskGeneration.go

```diff
@@ -0,0 +1,700 @@
+
+	return nil
+}
+
+// CreateDiskImage creates the final image by truncating the image with the proper size and
+// concatenating the contents of the given partitions.
+func (r *RawImage) CreateDiskImage(rawDiskFile string, partImgs []string) error {
+	var initDiskFile, endDiskFile string
+	var err error
+	var partFiles []string
+	var size uint64
+	var table partition.Table
+	var parts []*gpt.Partition
+
+	internal.Log.Logger.Debug().Str("disk", rawDiskFile).Strs("parts", partImgs).Msg("Creating disk image")
+
+	// Create disk image, 1Mb for alignment and GPT header, 2MB for bios boot partition
+	// Then concat all partition images
+	// Then add 1MB of free space at the end of the disk for gpt backup headers
+
+	// Create the start and end images
+	initDiskFile = filepath.Join(r.tmpDir, "init.raw")
+	endDiskFile = filepath.Join(r.tmpDir, "end.raw")
+	init, err := diskfs.Create(filepath.Join(r.tmpDir, "init.raw"), 3*1024*1024, diskfs.Raw, diskfs.SectorSizeDefault)
+	if err != nil {
+		internal.Log.Logger.Error().Err(err).Str("target", initDiskFile).Msg("failed to create init disk")
+		return err
+	}
+	end, err := diskfs.Create(filepath.Join(r.tmpDir, "end.raw"), 1*1024*1024, diskfs.Raw, diskfs.SectorSize512)
+	if err != nil {
+		internal.Log.Logger.Error().Err(err).Str("target", endDiskFile).Msg("failed to create end disk")
+		return err
+	}
+	err = init.Close()
+	if err != nil {
+		internal.Log.Logger.Error().Err(err).Str("target", initDiskFile).Msg("failed to close init disk")
+		return err
+	}
+	err = end.Close()
+	if err != nil {
+		internal.Log.Logger.Error().Err(err).Str("target", endDiskFile).Msg("failed to close end disk")
+		return err
+	}
+
+	// List and concatenate all image files
+	partFiles = append(partFiles, initDiskFile)
+	for _, img := range partImgs {
+		partFiles = append(partFiles, img)
+	}
+	partFiles = append(partFiles, endDiskFile)
+	err = utils.ConcatFiles(vfs.OSFS, partFiles, rawDiskFile)
+	if err != nil {
+		return err
+	}
+
+	// Add the partition table
+	finalDisk, err := diskfs.Open(rawDiskFile)
+	if err != nil {
+		internal.Log.Logger.Error().Err(err).Str("target", rawDiskFile).Msg("failed to open final disk")
+		return err
+	}
+	defer finalDisk.Close()
+
+	// Truncate file to multiple of sector size
+	stat, _ := os.Stat(rawDiskFile)
+	size = roundToNearestSector(stat.Size(), finalDisk.LogicalBlocksize)
+	err = os.Truncate(rawDiskFile, int64(size))
+	if err != nil {
+		internal.Log.Logger.Error().Err(err).Str("target", rawDiskFile).Msg("failed to truncate final disk")
+		return err
+	}
+
+	// Create a GPT partition table
+	// Size needs to be rounded to the nearest 512 bytes as that's the sector size
+	// Leave 1MB at the start for alignment+partition table (so start at sector 2048)
+	// 2MB for the BIOS boot partition in case we want to do hybrid boot
+	// Which would mean installing grub to the BIOS boot partition
+	// BIOS BOOT Legacy, currently does nothing
```

Comment:
omg! Good job figuring all this out!

Reply:
This and the code below took part of whats left from my sanity. There has to be a better way of building freaking partition tables in 2024.